include/merlin/memory_pool.cuh

```diff
@@ -0,0 +1,600 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime_api.h>
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+#include "debug.hpp"
+
+/**
+ * Allocators are used by the memory pool (and maybe other classes) to create
+ * RAII complient containers for buffers allocated in different memory areas.
+ */
+template <class T, class Allocator>
+struct AllocatorBase {
+  using type = T;
+  using sync_unique_ptr = std::unique_ptr<type, Allocator>;
+  using async_unique_ptr = std::unique_ptr<type, std::function<void(type*)>>;
+  using shared_ptr = std::shared_ptr<type>;
+
+  inline static sync_unique_ptr make_unique(size_t n) {
+    return sync_unique_ptr(Allocator::alloc(n));
+  }
+
+  inline static async_unique_ptr make_unique(size_t n, cudaStream_t ctx) {
+    return {Allocator::alloc(n, ctx), [ctx](type* p) { Allocator::free(p); }};
+  }
+
+  inline static shared_ptr make_shared(size_t n, cudaStream_t ctx = nullptr) {
+    return {Allocator::alloc(n, ctx),
+            [ctx](type* p) { Allocator::free(p, ctx); }};
+  }
+
+  inline void operator()(type* ptr) { Allocator::free(ptr, nullptr); }
+};
+
+/**
+ * Trivial fallback implementation using the standard C++ allocator. This mostly
+ * exists to ensure interface correctness, and as an illustration of what a
+ * proper allocator implementation should look like.
+ */
+template <class T>
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    delete[] ptr;
+  }
+};
+
+/**
+ * Claim/release buffers in pinned host memory.
+ */
+template <class T>
+struct HostAllocator final : AllocatorBase<T, HostAllocator<T>> {
+  using type = typename AllocatorBase<T, HostAllocator<T>>::type;
+
+  static constexpr const char* name{"HostAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
```

Comment:
Better to use `stream` for `cudaStream` var, as you know, there's really a context concept in CUDA world. https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html

Reply:
The type is `stream` the function is `context`. Hence the stream functions as the context in which the function executed.

Reply:
Hence the reason why it is called like that everywhere.